docs/programmers_guide/witness_formal_spec.md

```diff
@@ -2,100 +2,263 @@
+`nil` - an empty value.
 
-Each witness is a queue of instructions.
+`Any` - any data type. MUST NOT be `nil`.
 
-In every execution cycle a single instruction gets dequeued and a matching substitution rule gets applied to the stack.
+`Int` - an integer value. We treat the domain of integers as infinite,
+the overflow behaviour or mapping to the actual data types is undefined
+in this spec and should be up to implementation.
 
-In the end, when there are no more instructions left, there MUST be only one item left on the stack.  
+`Hash` - 32 byte value, representing a result of Keccak256 hashing.
 
-### Nodes
-
-There are multiple types of nodes that can be present on the stack.
+`ByteArray` - a byte array of arbitrary size. MUST NOT be empty.
 
-- `hashNode`
-- `branchNode`
-- `accountNode`
-- `valueNode`
-- `extensionNode`
+`()` - an empty array of arbitrary type.
 
-## Substitution rules
+`(Type1 Type1 Type1)` - an array of a type `Type1`.
 
-Here is an example substitution rule. The current instruction is named `INSTR1`.
+### Nodes
 
 ```
-STACK(node1, hash1) STACK(node2, hash2) INSTR1(params) |=> 
-STACK(helper1(node1, node2), helper2(hash1, hash2))
+type Node = HashNode{raw_hash:Hash}
+          | ValueNode{raw_value:ByteArray}
+          | AccountNode{nonce:Int balance:Int storage:nil|Node code:nil|CodeNode|HashNode}
+          | LeafNode{key:ByteArray value:ValueNode|AccountNode}
+          | ExtensionNode{key:ByteArray child:Node}
+          | BranchNode{child0:nil|Node child1:nil|Node child3:nil|Node ... child15:nil|Node}
+          | CodeNode{code:ByteArray}
 ```
 
-This substitution rule replaces 2 stack items `(node1, hash1)` and `(node2, hash2)`
-with a single stack item `(helper1(node1, node2), helper2(hash1, hash2))`.
+### Witness
 
-Where `helper1` and `helper2` are pure functions defined in pseudocode (see the example below).
+Witness MUST have at least 1 element.
 
 ```
-helper1 (value1, value2) {
-    return value1 + value2
+type WitnessHeader = {version:Int}
+type Witness = (Node|Instruction{code:Int parameter:Any...} ...)
+```
+
+
+## Execution Enviroment
+
+The witness execution environment MUST contain the following 2 elements:
+
+- **WitnessHeader** -- a header containing the version of the witness. The `version` MUST be 1.
+
+- **Witness** -- a witness to be executed;
+
+- **Substitution Rules** -- a list of all possible substitution rules.
+
+
+## Execution process
+
+Initially, the Witness MUST BE an array of `Instruction`s.
+
+Then, as substitution rules are applied to the witness, some elements of the
+array are replaces with `Node`s.
+
+The execution continues until there are no substitution rules left to execute.
+
+Here is how the execution code might look like in Go for building a single trie.
+
+```go
+
+witness := GetInitialWitness()
+rules := GetSubstitutionRules()
+numberOfRulesApplied := 1 // initial state
+
+for rulesApplied {
```

Comment:
What is this rulesApplied? Should probably be
for numberOfRulesApplied>0 {

Reply:
whoops, last minute rename :-D